### DISCUSSION TITLE
Enabling Analysis on Sources: Focus on Light Curve Fitting

### DISCUSSION THREAD

**Discussion by profjsb** [skyportal/skyportal#2638] [Ideas]
We plan to enable several different types of analysis on `Source`s, such as light curve fitting, spectral fits, and host galaxy typing. Much of this will be surfaced by custom per SkyPortal instance; such analysis may be computationally expensive and potentially rely upon proprietary code. In the this epic we intend to lay the ground work for enabling generic analysis with a focus on building in a light curve fitting using Type Ia supernovae fitter (via `sncosmo`) as an exemplar. It also starts the use of webhooks. Below maps the stories as park of this epoch

## Database Models

1. Create `AnalysisServices` table
    - `URL` to service (e.g. localhost)
    - `auth_token`
    - `auth_uses_header` bool
    - `enabled` = True
    - `name` (unique to a given SP instance)
    - `version` (semantic versioning or githash, for reference)
    - `contact_name` name of person responsible for the service (does not need to be part of this SP instance)
    - `contact_email` email address of contact person
    - `provides` one of the `analysis_types` enum (e.g. `lightcurve_fitting`)
    - `display_name`
    - `description`
    - `requires_data_types` a list of least one of `allowed_analysis_data_types` enum (e.g., `photometry` or `spectra`) - this is what will be sent over to the service
    - `timeout_limit` 3600 - maximum allowed time to return results or the webhook is invalidated
    - `allowed_groups` = “all”  (or a list of ids) - who can use this service

2. Create `Webhook` table
    - `created_by`
    - `created_at`
    - `invalid_after`
    - `token` unique, long-string token for the callback
    - `status` ([ `timed_out` | `started` | `completed`)
    - `handled_by`: url for internal API to handle the incoming result

3. Create `Analysis` table 
    - `AnalysisServices`_id
    - `Webhook`_id
    - `Obj`_id
    - `group_ids` - who can see the results
    - `duration` - how long did this take to run and return
    - `results` - JSONB of results (parameters, plots, inferencedata)    
        [{“parameters”: {”t_max_mjd”: {”val”: 564921.2, “err”: 0.02}, “chi2/dof”: 1.045, ...},
         “plot”: {”name”: “ztf_r_fit”, “data”: b”23a34bc...”, “caption”: None},
         “inferenceData”: binary representation of an arviz.InferenceData type},
        ]
    - `show_parameters` : list of parameters to show in a table
    - `show_plots` : list of plots to show
    - `show_corner`: list of variables to show in a corner.corner plot of `results_inferencedata`  (empty if do not show)
    - `active` True

## APIs

 4.  Create `AnalysisServices` APIs

   - admin permissions for POST, DELETE, PATCH
   - GET (all active or by id)

5. Create `Webhook` APIs
    - POST, DELETE, GET
    - invalidate a POST if it comes too late

6. Create `Analysis` APIs
    - POST, DELETE, GET (all or by Id)
    

## Front end

7. Show Analysis Results Page by Source (/source/<id>/analysis)
   - Show all active Analysis results viewable by user
   - Stub for generate analysis section
   - delete button for analysis id

8. Start Button Analysis
   - select from the available analysis
   - package up available data and send (netcdf file)?

## Middleware

 9. sncosmo microservice

1. Connect analysis initiation with sncosmo webhook

**Comment by guynir42** [comment]
Minor suggestions: 
- `Analysis` -> `AnalysisResults`
- We could inherit from `AnnotationsMixin` to have the same author, groups logic as annotations.
- We should not have `AnalysisResult` be tied to an `obj_id` by default, or at least have an `AnalysisMixin` and then inherit from that because some analysis results would be on a source, but some would be on a spectrum, some on a GCN, etc. 
Just some lessons learned from comments/annotations.

**Comment by stefanv** [comment]
I am not clear to me what the role of the webhooks will be here. Can results not be posted directly back to the analysis API with the service's registered token?

**Reply by guynir42** [reply]
If I send a request to do some processing on a source, how would the front-end know that a result has been posted?

**Reply by profjsb** [reply]
I think we need to generate a call-specific token so that the generator of the results does not need to use the auth of the person who generated the token. Also, perhaps more important, there's generic webhook capabilities (like time invalidation, cancel webhooks, webhook statistics) that I think we want to build in. Makes sense?

**Reply by profjsb** [reply]
> If I send a request to do some processing on a source, how would the front-end know that a result has been posted?

When the analysis is done and loaded that API can do a refresh sources to push the updates to the front end.

**Comment by profjsb** [comment]
Update: the backend/API & middleware is now in place. We're going to focus on frontend now.

**Comment by profjsb** [comment]
Ideas for the front-end development (order):

0) Dedicated Analysis Service page (#3193)
    - Admins can add new services on the front end, delete, etc.
    - Users can see available services to them (including input parameters)
    
1) Show analysis by source
  - a page to show the results of a specific analysis. Endpoint something like `/source/<source_id>/analysis/<analysis_number>`
  - shows plots, high-level results, dowload buttons

2) Source-page analysis panel
   - show a list of all analyses 
   - dont show failed analyses by default
   - drop down menu to start a new analysis
